joblib/numpy_pickle_utils.py

```diff
@@ -0,0 +1,558 @@
+_ZLIB_PREFIX = b'\x78'
+_GZIP_PREFIX = b'\x1f\x8b'
+_BZ2_PREFIX = asbytes('BZ')
+_XZ_PREFIX = b'\xfd\x37\x7a\x58\x5a'
+_LZMA_PREFIX = b'\x5d\x00'
+
+# Supported compressors
+_COMPRESSORS = ('zlib', 'bz2', 'lzma', 'xz', 'gzip')
+
+# The max magic number length of supported compression file types.
+_MAX_PREFIX_LEN = len(max((_ZFILE_PREFIX, _GZIP_PREFIX, _BZ2_PREFIX,
+                           _XZ_PREFIX, _LZMA_PREFIX,), key=len))
+
+# Buffer size used in io.BufferedReader and io.BufferedWriter
+_IO_BUFFER_SIZE = 10 * 1024 ** 2
+
+
+###############################################################################
+# Cache file utilities
+def _first_bytes(filename):
+    """Return the first _MAX_PREFIX_LEN bytes of a filename.
+
+    Parameters
+    ----------
+    filename: str
+    """
+    with open(filename, 'rb') as f:
+        first_bytes = f.read(_MAX_PREFIX_LEN)
+
+    return first_bytes
+
+
+def _buffered_read_file(fobj):
+    """Return a buffered version of a read file object."""
+    if PY26 or (PY27 and isinstance(fobj, bz2.BZ2File)):
+        # Python 2.6 doesn't fully support io.BufferedReader.
+        # Python 2.7 doesn't work with BZ2File through a buffer: "no
+        # attribute 'readable'" error.
+        # In those cases, we directly return the file object.
+        return fobj
+    else:
+        return io.BufferedReader(fobj, buffer_size=_IO_BUFFER_SIZE)
+
+
+def _buffered_write_file(fobj):
+    """Return a buffered version of a write file object."""
+    if PY26 or (PY27 and isinstance(fobj, bz2.BZ2File)):
+        # Python 2.6 doesn't fully support io.BufferedWriter.
+        # Python 2.7 doesn't work with BZ2File through a buffer: no attribute
+        # 'writable'.
+        # In those cases, we directly return the file object.
+        return fobj
+    else:
+        return io.BufferedWriter(fobj, buffer_size=_IO_BUFFER_SIZE)
+
+
+def _read_fileobject(filename):
+    """Utility function opening the right fileobject from a filename.
+
+    The magic number is used to choose between the type of file object to open:
+    * regular file object (default)
+    * zlib file object if magic matches _ZLIB_PREFIX
+    * gzip file object if magic matches _GZIP_PREFIX or_GZIP_PREFIX_2
+    * bz2 file object if magic matches _BZ2_PREFIX
+    * lzma file object if magic matches _LZMA_PREFIX or _XZ_PREFIX
+    or _LZMA_PREFIX_3
+
+    Parameters
+    ----------
+    filename: str
+        The full path string of the file.
+
+    Returns
+    -------
+        a file like object
+
+    """
+    first_bytes = _first_bytes(filename)
+    if first_bytes.startswith(_ZLIB_PREFIX):
+        return _buffered_read_file(BinaryZlibFile(filename, 'rb'))
+    elif first_bytes.startswith(_GZIP_PREFIX):
+        return _buffered_read_file(BinaryGzipFile(filename, 'rb'))
+    elif first_bytes.startswith(_BZ2_PREFIX):
+        return _buffered_read_file(bz2.BZ2File(filename, 'rb'))
+    elif any(map(lambda each: first_bytes.startswith(each),
```

Comment:
Too clever I would say:

```python
if first_bytes.startswith(_LZMA_PREFIX) or first_bytes(startswith(_XZ_PREFIX)):
```

If you insist, rather than a map + lambda, comprehensions is a lot better suited IMHO:
```python
any(first_bytes.startswith(prefix) for prefix in (_LZMA_PREFIX, _XZ_PREFIX))
```